add Demos/Wait-for-Acknowledgment. (#5184)
ros2/ros2_documentation: source/Tutorials/Demos/Wait-for-Acknowledgment.rst

> Wait for acknowledgement
> ========================

Reviewer: To be consistent, and since I think we should probably use "acknowledgment" ( :us: ) instead of "acknowledgement" ( :gb: ):

```suggestion
Wait for acknowledgment
=======================
```

**Goal:** Wait for acknowledgment of messages sent by a publisher.

**Tutorial level:** Advanced

**Time:** 10 minutes

.. contents:: Table of Contents
   :depth: 1
   :local:

Overview
--------

In Publisher-Subscriber architecture, messages are sent from the publisher to the subscribers, and the publisher does not have any built-in mechanism to confirm that the subscriber has received the messages.

> This feature enables the publisher to wait for acknowledgment of messages sent by the publisher.

Reviewer: ```suggestion
This feature enables the publisher to wait for acknowledgment of messages it sent.
```

This is useful in scenarios where the publisher needs to ensure that the subscriber has received the message before proceeding with further actions, such as sending more messages or performing other operations.

RMW Support
-----------

> Wait for acknowledgement requires RMW implementation support.

Reviewer: ```suggestion
Wait for acknowledgment requires RMW implementation support.
```

.. list-table::  Wait-for-Acknowledgment Support Status
   :widths: 25 25

   * - rmw_fastrtps
     - supported
   * - rmw_connextdds
     - supported
   * - rmw_cyclonedds
     - supported

Installing the demo
-------------------

See the :doc:`installation instructions <../../Installation>` for details on installing ROS 2.

If you've installed ROS 2 from packages, ensure that you have ``ros-{DISTRO}-examples-rclcpp-minimal-publisher`` and ``ros-{DISTRO}-examples-rclcpp-minimal-subscriber`` installed.
If you downloaded the archive or built ROS 2 from source, it will already be part of the installation.

Running the demo
----------------

> This demo shows how to use the wait for acknowledgment feature in the publisher to ensure that messages sent by the publisher are acknowledged by the subscriber.

Reviewer: ```suggestion
This demo shows how to use the wait for acknowledgment feature in the publisher to ensure that messages sent by the publisher are acknowledged by all subscriptions.
```

I think this is correct, right?

Author: yes this is precisely true.

> https://github.com/ros2/examples/blob/{REPOS_FILE_BRANCH}/rclcpp/topics/minimal_publisher/member_function_with_wait_for_all_acked.cpp

Reviewer: Since we're including the whole file here, could we instead just link to the file without copying its contents into the article? This way we don't have to update this if the source code file changes. I think we do include whole files in other demos (e.g., the content filtering demo), but just a thought. This is more of a potential long-term improvement.

Author: yeah i like that. i just followed the other tutorial documentation like content filtering. maybe that would be best for us to maintain the demo code and documentation. let's do that. (note that i want to keep this commit to address the original issue, so other parts are going to be fixed with another PRs.)

          RCLCPP_INFO(
            this->get_logger(),
            "All subscribers acknowledge messages");
        } else {
          RCLCPP_INFO(
            this->get_logger(),
            "Not all subscribers acknowledge messages during %" PRId64 " ms",
            static_cast<int64_t>(wait_timeout_.count()));
        }
      }

      void timer_callback()
      {
        auto message = std_msgs::msg::String();
        message.data = "Hello, world! " + std::to_string(count_++);
        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
        publisher_->publish(message);

        // After sending some messages, you can call wait_for_all_acked() to confirm all subscribers
        // acknowledge messages.
      }
      rclcpp::TimerBase::SharedPtr timer_;
      rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
      size_t count_;
      std::chrono::milliseconds wait_timeout_;
    };

    int main(int argc, char * argv[])
    {
      rclcpp::init(argc, argv);

      auto publisher = std::make_shared<MinimalPublisher>();
      rclcpp::spin(publisher);
      rclcpp::shutdown();

      return 0;
    }

> The publisher can use the ``wait_for_all_acked`` method to wait for message acknowledgments within a specified timeout before shutdown by the signal.

Reviewer: Would it be worth explicitly pointing out here in the text that the publisher's reliability QoS needs to be set to reliable, even if it's mentioned in a comment in the code? I'll let you decide.

https://github.com/ros2/rclcpp/blob/7b6ee8a2e7a13d73f9f69368970390a9e0930448/rclcpp/include/rclcpp/publisher_base.hpp#L247-L248

Author: good catch, this is very important information that user needs to know. i added this note in the `RMW support` section.

> We can start the demo by running the ``ros2 run examples_rclcpp_minimal_publisher publisher_wait_for_all_acked`` and ``ros2 run examples_rclcpp_minimal_subscriber subscriber_member_function`` executable (don't forget to source the setup file first):

Reviewer: Minor suggestion:

```suggestion
We can start the demo by running the ``publisher_wait_for_all_acked`` and ``subscriber_member_function`` executable from the ``examples_rclcpp_minimal_publisher`` package (don't forget to source the setup file first):
```

Start the subscriber in one terminal:

> .. code-block:: bash

Reviewer: ```suggestion
.. code-block:: console
```

Author: Ah, sorry. complete command for `bash`, and terminal output to show for `console`...

    $ ros2 run examples_rclcpp_minimal_subscriber subscriber_member_function
    [INFO] [1743121567.030751270] [minimal_subscriber]: I heard: 'Hello, world! 0'
    [INFO] [1743121567.530981660] [minimal_subscriber]: I heard: 'Hello, world! 1'
    [INFO] [1743121568.031032935] [minimal_subscriber]: I heard: 'Hello, world! 2'
    [INFO] [1743121568.531048458] [minimal_subscriber]: I heard: 'Hello, world! 3'
    [INFO] [1743121569.031049351] [minimal_subscriber]: I heard: 'Hello, world! 4'
    [INFO] [1743121569.530980327] [minimal_subscriber]: I heard: 'Hello, world! 5'
    [INFO] [1743121570.030825871] [minimal_subscriber]: I heard: 'Hello, world! 6'
    ...

Then start the publisher in another terminal:

> .. code-block:: bash

Reviewer: ```suggestion
.. code-block:: console
```

    $ ros2 run examples_rclcpp_minimal_publisher publisher_wait_for_all_acked
    [INFO] [1743121567.030353553] [minimal_publisher_with_wait_for_all_acked]: Publishing: 'Hello, world! 0'
    [INFO] [1743121567.530420788] [minimal_publisher_with_wait_for_all_acked]: Publishing: 'Hello, world! 1'
    [INFO] [1743121568.030461599] [minimal_publisher_with_wait_for_all_acked]: Publishing: 'Hello, world! 2'
    [INFO] [1743121568.530435646] [minimal_publisher_with_wait_for_all_acked]: Publishing: 'Hello, world! 3'
    [INFO] [1743121569.030431263] [minimal_publisher_with_wait_for_all_acked]: Publishing: 'Hello, world! 4'
    [INFO] [1743121569.530447106] [minimal_publisher_with_wait_for_all_acked]: Publishing: 'Hello, world! 5'
    [INFO] [1743121570.030353934] [minimal_publisher_with_wait_for_all_acked]: Publishing: 'Hello, world! 6'
    ^C[INFO] [1743121570.344981639] [rclcpp]: signal_handler(signum=2)
    [INFO] [1743121570.345398788] [minimal_publisher_with_wait_for_all_acked]: All subscribers acknowledge messages

> When the publisher is terminated (e.g., by pressing Ctrl+C), it will wait for acknowledgment of all messages sent before shutdown.

Reviewer: We can make it fancy! https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-kbd

![image](https://github.com/user-attachments/assets/b0f41496-59bd-4166-b549-723592530de4)

```suggestion
When the publisher is terminated (e.g., by pressing :kbd:`Ctrl-C`), it will wait for acknowledgment of all messages sent before shutdown.
```

Author: Nicer, thanks for the tip. I did not know this.

Wait for acknowledgment
=======================

**Goal:** Wait for acknowledgment of messages sent by a publisher.

**Tutorial level:** Advanced

**Time:** 10 minutes

.. contents:: Table of Contents
   :depth: 1
   :local:

Overview
--------

In Publisher-Subscriber architecture, messages are sent from the publisher to the subscribers, and the publisher does not have any built-in mechanism to confirm that the subscriber has received the messages.
This feature enables the publisher to wait for acknowledgment of messages it sent.
This is useful in scenarios where the publisher needs to ensure that the subscriber has received the message before proceeding with further actions, such as sending more messages or performing other operations.

RMW Support
-----------

Wait for acknowledgment requires RMW implementation support.

.. list-table::  Wait-for-Acknowledgment Support Status
   :widths: 25 25

   * - rmw_fastrtps
     - supported
   * - rmw_connextdds
     - supported
   * - rmw_cyclonedds
     - supported

.. note::

>    The publisher's QoS profile needs to be ``RELIABLE`` to use the wait for acknowledgment feature, otherwise the publisher will not wait for acknowledgment.

Reviewer: Just to be clearer and link to more information:

```suggestion
   The publisher's :ref:`QoS reliability policy <about_qos_policies>` needs to be ``RELIABLE`` to use the wait for acknowledgment feature, otherwise the publisher will not wait for acknowledgment.
```

With the corresponding target:

```diff
diff --git a/source/Concepts/Intermediate/About-Quality-of-Service-Settings.rst b/source/Concepts/Intermediate/About-Quality-of-Service-Settings.rst
index 36962b89..7d24436c 100644
--- a/source/Concepts/Intermediate/About-Quality-of-Service-Settings.rst
+++ b/source/Concepts/Intermediate/About-Quality-of-Service-Settings.rst
@@ -24,6 +24,8 @@ QoS profiles can be specified for publishers, subscriptions, service servers and
 A QoS profile can be applied independently to each instance of the aforementioned entities, but if different profiles are used, it is possible that they will be incompatible, preventing the delivery of messages.
 
 
+.. _about_qos_policies:
+
 QoS policies
 ------------

```

Reviewer: or maybe it should link to this section with the compatibility table? https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Quality-of-Service-Settings.html#qos-compatibilities

Author: i will borrow the former one, but `:ref:` can not be rendered in `note::` section, so i just removed `note::` section.

Reviewer: It worked for me, but this is also fine

Installing the demo
-------------------

See the :doc:`installation instructions <../../Installation>` for details on installing ROS 2.

If you've installed ROS 2 from packages, ensure that you have ``ros-{DISTRO}-examples-rclcpp-minimal-publisher`` and ``ros-{DISTRO}-examples-rclcpp-minimal-subscriber`` installed.
If you downloaded the archive or built ROS 2 from source, it will already be part of the installation.

Running the demo
----------------

This demo shows how to use the wait for acknowledgment feature in the publisher to ensure that messages sent by the publisher are acknowledged by all subscriptions.

https://github.com/ros2/examples/blob/{REPOS_FILE_BRANCH}/rclcpp/topics/minimal_publisher/member_function_with_wait_for_all_acked.cpp

The publisher can use the ``wait_for_all_acked`` method to wait for message acknowledgments within a specified timeout before shutdown by the signal.

> We can start the demo by running the ``publisher_wait_for_all_acked`` and ``subscriber_member_function`` executable from the ``examples_rclcpp_minimal_publisher`` package (don't forget to source the setup file first):

Reviewer: My bad:

```suggestion
We can start the demo by running the ``publisher_wait_for_all_acked`` and ``subscriber_member_function`` executables from the ``examples_rclcpp_minimal_publisher`` package (don't forget to source the setup file first):
```
